python/cugraph-equivariant/cugraph_equivariant/nn/tensor_product_conv.py

```diff
@@ -0,0 +1,252 @@
+    function. For example, they can be constructed from edge embeddings and
+    scalar features:
+
+    .. math::
+        \psi_{a b} = \operatorname{MLP}
+        \left(e_{a b}, \mathbf{h}_a^0, \mathbf{h}_b^0\right)
+
+    Parameters
+    ----------
+    in_irreps : e3nn.o3.Irreps
+        Irreps for the input node features.
+
+    sh_irreps : e3nn.o3.Irreps
+        Irreps for the spherical harmonic representations of edge vectors.
+
+    out_irreps : e3nn.o3.Irreps
+        Irreps for the output.
+
+    batch_norm : bool, optional (default=True)
+        If true, batch normalization is applied.
+
+    mlp_channels : sequence of ints, optional (default=None)
+        A sequence of integers defining number of neurons in each layer in MLP
+        before the output layer. If `None`, no MLP will be added.
+
+    mlp_activation : callable, optional (default=torch.nn.GELU)
+        Activation function in MLP.
+
+    mlp_fast_first_layer : bool, optional (default=False)
+        When enabled, for the first layer of MLP, the module performs GEMMs on
+        individual components (i.e., edge embeddings, scalar features of source
+        and destinations nodes) of the weights before indexing and concatenation,
+        leading to a lower complexity in most use cases. This option requires
+        users to explicitly pass in `src_scalars` and `dst_scalars` in
+        `forward()` call.
+
+    use_e3nn_tp: bool, optional (default=False)
+        If `True`, use TensorProduct functions from e3nn.
+
+    Examples
+    --------
+    >>> # Case 1: MLP with the input layer having 6 channels and 2 hidden layers
+    >>> #         having 16 channels. edge_emb.size(1) must match the size of
+    >>> #         the input layer: 6
+    >>>
+    >>> conv1 = FullyConnectedTensorProductConv(in_irreps, sh_irreps, out_irreps,
+    >>>     mlp_channels=[6, 16, 16], mlp_activation=nn.ReLU).cuda()
+    >>> out = conv1(src_features, edge_sh, edge_emb, graph)
+    >>>
+    >>> # Case 2: No MLP, edge_emb will be directly used as the tensor product weights
+    >>>
+    >>> conv2 = FullyConnectedTensorProductConv(in_irreps, sh_irreps, out_irreps,
+    >>>     mlp_channels=None).cuda()
+    >>> out = conv2(src_features, edge_sh, edge_emb, graph)
+    >>>
+    >>> # Case 3: Same as case 1 but with `mlp_fast_first_layer=True`. The scalar
+    >>> #         features from edges, sources and destinations have to be passed
+    >>> #         in separately.
+    >>>
+    >>> conv3 = FullyConnectedTensorProductConv(in_irreps, sh_irreps, out_irreps,
+    >>>     mlp_channels=[6, 16, 16], mlp_fast_first_layer=True).cuda()
+    >>> out = conv3(src_features, edge_sh, edge_scalars, graph,
+    >>>     src_scalars=src_scalars, dst_scalars=dst_scalars)
+    """
+
+    def __init__(
+        self,
+        in_irreps: o3.Irreps,
+        sh_irreps: o3.Irreps,
+        out_irreps: o3.Irreps,
+        batch_norm: bool = True,
```

Comment:
In the Diffdock training code, we have customized batch_norm function. Instead of being a boolean option, can this take a callable that defaults to `e3nn.nn.BatchNorm`?

Reply:
Sure. Will make the change.

Reply:
I haven't came up with a neat way to support this because if we change the argument to be a callable (with the default value `=Batchnorm`), the customized batch_norm function must have the same function signature as e3nn.nn.Batchnorm. With that, I would suggest to apply customized function outside of the conv layer.

Reply:
Both the e3nn.nn.BatchNorm and the modified version we used are nn.module so I should have suggested change the type of the `batch_norm` argument to `nn.module`. But then it becomes a moot point since we can apply the BatchNorm outside of the conv layer like you said. Let me double check with Guoqing, who authored the modified version of BatchNorm

Reply:
Just confirmed with Guoqing that the current API is OK. So it's up to you if you want to keep it as it is

Reply:
Wouldn't it make more sense to just not have BN in the API then?